src/Data/BitVector/Sized/Internal.hs

```diff
@@ -1,419 +1,834 @@
+-- refers to an actual bit in the input 'BV'. If it is out of range,
+-- just silently return the original input.
+setBit' :: NatRepr w
+        -- ^ Desired output width
+        -> Natural
+        -- ^ Index of bit to set
+        -> BV w
+        -- ^ Original bitvector
+        -> BV w
+setBit' w ix bv
+  | ix < natValue w = or bv (BV (B.bit (fromNatural ix)))
+  | otherwise = bv
+
+-- | @clearBit w bv ix@ is the same as @and bv (complement (bit w ix))@.
+clearBit :: ix+1 <= w
+         => NatRepr w
+         -- ^ Desired output width
+         -> NatRepr ix
+         -- ^ Index of bit to clear
+         -> BV w
+         -- ^ Original bitvector
+         -> BV w
+clearBit w ix bv =
+  -- NB, fromNatural is OK because of the (ix+1<=w) constraint
+  and bv (complement w (BV (B.bit (fromNatural (natValue ix)))))
+
+
+-- | Like 'clearBit', but without the requirement that the index bit
+-- refers to an actual bit in the input 'BV'. If it is out of range,
+-- just silently return the original input.
+clearBit' :: NatRepr w
+          -- ^ Desired output width
+          -> Natural
+          -- ^ Index of bit to clear
+          -> BV w
+          -- ^ Original bitvector
+          -> BV w
+clearBit' w ix bv
+  | ix < natValue w = and bv (complement w (BV (B.bit (fromNatural ix))))
+  | otherwise = bv
+
+-- | @complementBit bv ix@ is the same as @xor bv (bit knownNat ix)@.
+complementBit :: ix+1 <= w
+              => NatRepr ix
+              -- ^ Index of bit to flip
+              -> BV w
+              -- ^ Original bitvector
+              -> BV w
+complementBit ix bv =
+  -- NB, fromNatural is OK because of (ix+1 <= w) constraint
+  xor bv (BV (B.bit (fromNatural (natValue ix))))
+
+-- | Like 'complementBit', but without the requirement that the index
+-- bit refers to an actual bit in the input 'BV'. If it is out of
+-- range, just silently return the original input.
+complementBit' :: NatRepr w
+               -- ^ Desired output width
+               -> Natural
+               -- ^ Index of bit to flip
+               -> BV w
+               -- ^ Original bitvector
+               -> BV w
+complementBit' w ix bv
+  | ix < natValue w = xor bv (BV (B.bit (fromNatural ix)))
+  | otherwise = bv
 
 -- | Test if a particular bit is set.
-bvTestBit :: BitVector w -> Int -> Bool
-bvTestBit (BV _ x) b = testBit x b
-
--- | Get the number of 1 bits in a 'BitVector'.
-bvPopCount :: BitVector w -> Int
-bvPopCount (BV _ x) = popCount x
-
--- | Truncate a bit vector to a particular width given at runtime, while keeping the
--- type-level width constant.
-bvTruncBits :: BitVector w -> Int -> BitVector w
-bvTruncBits (BV wRepr x) b = BV wRepr (truncBits b x)
+testBit :: ix+1 <= w => NatRepr ix -> BV w -> Bool
+testBit ix (BV x) = B.testBit x (fromNatural (natValue ix))
+
+-- | Like 'testBit', but takes a 'Natural' for the bit index. If the
+-- index is out of bounds, return 'False'.
+testBit' :: Natural -> BV w -> Bool
```

Comment:
This will panic instead of returning false on too large index values. (might be my fault, oops)

Reply:
Fixed